feat: replace current match

stefanwatt opened the pull request:
## Checklist
- [x] Used only camelCase variable names.
- [ ] If functionality is added or modified, also made respective changes to the
  `README.md` (the `.txt` file is auto-generated and does not need to be modified).

This is just a rough first draft. There's TODOs and bugs still, but I would like to get some feedback at this point to make sure I'm going in a direction that aligns with the project. I tried my best to stick with your naming scheme etc..
I was itching to do some more refactoring, but I want to be respectful to your project. Maybe we can discuss that at some point.

I have not yet adapted the readme, because I wanted to make sure the suggested changes are ok with you.

added functionality:

- keep list of matches in state
- keep selected match in state
- add keymaps for cycling through matches
- adapt keymaps for confirmation -> confirm/confirmAll
- updated cursor position in targetWin and re-center if necessary
- confirm will only replace the selected match

known issues:
- seems like something went wrong when solving conflicts
- replacement preview shows an extra character that should not be shown
- replacement preview is sluggish when there's a lot of matches

chrisgrieser commented:
Hmm, yes, this appears be a larger PR than I expected…

I don't mind refactorings, but they should be a separate PR for clarity reasons. I think it might sense to split up this feature into separate PRs maybe? Like one big PR for all the refactoring you need in preparation for the feature, and then another PR for the feature itself. That way, you can get changes already in, and it would also be easier to review the PR

stefanwatt commented:
Perhaps that wasn't clear, but I did not do any optional refactoring, yet.

> Like one big PR for all the refactoring you need in preparation for the feature, and then another PR for the feature itself

Sorry but I have no idea what you mean by this or how that could be done. It sounds to me like you are mainly perturbed by the amount of changes? I can't think of a way to split this that would make any sense. The changes I made are directly related to eachother. There's not really anything that can be handled separately, I don't think.

stefanwatt commented:
I mean we could probably do the changes to how matches are stored and then used later in a separate PR,
but that's like 95% of this PR.
Then the PR for this feature would be very small, but I'm not convinced that will improve clarity.

chrisgrieser commented:
So, the current PR changes a total of ~600 lines. Considering that the whole repo is only barely larger, this is unreasonably large. 

Looking into it, it appears that the PR has various types of changes:
1. cosmetic ones (like fixing the casing of classes)
2. Formatting ones, as it appears you are using something with with a shorter line length. This anime is responsible for half the diff. 
3. Refactorings that prepare the new feature, such as restructured settings or turning local functions into exported functions
4. The feature itself
5. Leftover log statements 

1. I don't mind, but it should be it's own PR
2. Should be fixed on your end by configuring your formatter properly, it creates so much noise the PR is hard to review
3 and 4 are the actual feature, which as far as I can see, could be split up to make the changes easier to follow. 
5. Should be removed

To distinguish 3 and 4, you could think about how far you can make "prepratory" changes without actually adding in the new feature. 

Also, your PR is based on an outdated version of the repo, which results in summer stuff like the regex101 config missing in your fork. You should sync up your fork. 

Like all these changes are not sufficient for the PR to be merged, but only for the PR to be in a state where it can be properly reviewed.

stefanwatt commented:
I understand this needs to be cleaned up and is not ready for review. Like I said, I wanted to synch with you on the direction before I continue. Sorry if this is not standard procedure. 

1. As far as I'm aware I changed the casing in one instance which seemed to clearly be a typo. You capitalized class names everywhere else. I'm not creating a separate PR for a typo.

2. I use stylua with default settings. Seems that print width is 120 chars by default. That seems pretty generous already. What do you use for formatting? Can we somehow synch our settings?

3. and 4. I can certainly try to split this up some, but I also don't want to put in work for that unless I know where I'm going. Perhaps you could give me some directions here. 

5. Totally fair. Of course that needs to be removed before it can be merged. I'm still planning on doing development and I want to continue to use the debug print statements.
I have now removed them for this PR tho and will stash/pop them as needed.

chrisgrieser commented on lua/rip-substitute/rg-operations.lua:
```diff
@@ -1,12 +1,33 @@
 local M = {}
+local config = require("rip-substitute.config").config
```
`config` should not be called at the top of the file, since it's then only set once. This prevents the user from running `.setup` again to change a config. Thus, `config` should always be set dynamically inside a function

stefanwatt replied:
I was wondering about that. For some reason I feel kinda dirty requiring it on every execution of a function.
Is there any drawback to the way you're suggesting?

chrisgrieser replied:
no, since lua cache's is anyway.

stefanwatt replied:
Ok great. That's an obvious choice then.

chrisgrieser commented on nvim-float.log:
leftover file

chrisgrieser commented on lua/rip-substitute/utils.lua:
```diff
@@ -1,6 +1,26 @@
+---@generic T
+---@param list Array<`T`>
+---@param cb function(value: `T`,i:integer): `T`
+M.map = function(list, cb)
```
1. For consistency with the rest of the repo, it should be `function M.map`
2. Is there really a need for a map-util, when nvim already offers us `vim.tbl_map` and `Iter:map()`? If so, the explanation why a custom map implementation is needed would be something that should be noted in a comment

stefanwatt replied:
using vim.tbl_map now

chrisgrieser commented on lua/rip-substitute/rg-operations.lua:
```diff
@@ -91,113 +115,131 @@ function M.executeSubstitution()
+function M.highlightReplacement(match, selected, targetBuf, incPreviewNs)
+	local hl_group = selected and config.incrementalPreview.hlGroups.currentMatch
+		or config.incrementalPreview.hlGroups.replacement
+	local ok, err = pcall(
```
`pcall` isn't ideal, I think. In my view, it should only be used to wrap functions that are expected to throw errors in normal usage (such as `require`, when the specified module does not exist). I think this is not the case for `nvim_buf_set_extmark`, afaik, the function does not throw an error when used correctly

stefanwatt replied:
true, I removed it

chrisgrieser commented on lua/rip-substitute/popup-win.lua:
```diff
@@ -310,6 +317,15 @@ function M.openSubstitutionPopup(searchPrefill)
 		group = vim.api.nvim_create_augroup("rip-substitute-popup-changes", {}),
 		callback = function()
 			ensureOnly2LinesInPopup()
+			local updatedMatches, err = rgMatches.getMatches()
+			if err then
+				vim.print(err)
+			else
+				state.matches = updatedMatches
+				--TODO: this should not always run
+				state.selectedMatch = rgMatches.getClosestMatchAfterCursor(state.matches)
+				if state.selectedMatch then rgMatches.centerViewportOnMatch(state.selectedMatch) end
+			end
```
yeah, I think this block should not be triggered by `TextChanged`, but only on running confirming/next/prev.

chrisgrieser commented on lua/rip-substitute/popup-win.lua:
```diff
@@ -67,11 +68,17 @@ end
 
 local function confirmSubstitution()
```
I think for clarity reasons, this should now be renamed to `confirmSingleSubstitution`

stefanwatt replied:
makes sense 👍 renamed it

chrisgrieser commented on lua/rip-substitute/matches.lua:
```diff
@@ -0,0 +1,221 @@
+			local rgArgs = {
+				"--line-number",
+				"--column",
+				"--vimgrep",
```
I think `--vimgrep` is just a shorthand that bundles a bunch of other options. I think it actually not needed, since the other options already cover it

stefanwatt replied:
removed 👍

chrisgrieser commented on lua/rip-substitute/matches.lua:
```diff
@@ -0,0 +1,221 @@
+---@param match RipSubstituteMatch
+function M.centerViewportOnMatch(match)
+	local state = require("rip-substitute.state").state
+	local padding = 5
```
any reason why the padding is 5? Would it maybe make sense to use the user's value for `vim.o.scrolloff` instead?

stefanwatt replied:
absolutely, great suggestion

chrisgrieser reviewed (commented):
Just some minor things I noticed while going through the code

chrisgrieser commented:
> Like I said, I wanted to synch with you on the direction before I continue. Sorry if this is not standard procedure.

tbh, I am also not really familiar what the "standard procedure" is. Nonetheless, even with just to get some feedback on the ongoing direction is hard to give, if there is so much noise in the PR, so the cleanup would be needed regardless

> As far as I'm aware I changed the casing in one instance which seemed to clearly be a typo. You capitalized class names everywhere else. I'm not creating a separate PR for a typo.

Yes, and due 600+ lines of diff, it is not easy to tell whether this is a one-time occurrence or something that happens multiple times. ;)

> I use stylua with default settings. Seems that print width is 120 chars by default. That seems pretty generous already. What do you use for formatting? Can we somehow synch our settings?
I ran the formatter again on all the files and that seems to have reduced the number of changes quite a bit.

Not sure what exactly you did, but before, the diff had a lot of cases where it split up one line into multiple lines, indicating that some formatter with lower line length was run. Anyway, I use stylua with the config provided in the repo. If you simply run stylua in the repo root, it should by default pick up the config by itself. At quick glance, the diff looks okay now.

Sorry, if I am came off as nitpicky or something, imho keeping noise to a minimum is essential in large PRs, as it's otherwise hard to assess what actual changes have been made. And even if this is a draft PR and therefore still WIP, a large PR with a lot of noise makes it harder to understand what's going on and thus also hard to even give intermediary feedback. In hindsight, most of the issue was due to the formatting issues, but that 1. (and 3.) are less of an issue was in itself hard to tell with so much noise going on there.

---

> 3. and 4. I can certainly try to split this up some, but I also don't want to put in work for that unless I know where I'm going. Perhaps you could give me some directions here. What constitutes a preparatory change vs a feature change? When do you find it appropriate to split something into a new PR vs when should it remain together?

Well, you do a lot of refactorings which make sense, but can be split off into its own PR. And keeping PRs as small is possible is usually desirable (better readability, thus easier reviewability, also cleaner git history).
- you do some function renaming, such as `executeSubstitution` to `substituteAll`, which make sense, but are only refactorings. 
- the casing change mentioned in 1 could also go here.
- many of the changes in the incremental-preview function of `rg-operations.lua` seem to be refactorings for making it easier for your new `rg-matches` module to hook in. These refactorings can be made independently of adding your feature. 

By doing so, the remaining changes in `rg-operations.lua` should only be those where your match-logic is inserted, thus making it more transparent how your new feature interacts with the existing plugin code. (Right now, the diff of `rg-operations` is hard to review, since a lot of it is just code moving around.)

chrisgrieser commented:
regarding the remaining issues:

> seems like something went wrong when solving conflicts

I *think* this is not the case anymore with your latest changes.

> replacement preview shows an extra character that should not be shown

That's probably just an off-by-one issue somewhere since some nvim functions expect columns are zero-based and some expect them to be one-based. 

> replacement preview is sluggish when there's a lot of matches

Currently (without your PR), `rg` is run exactly twice on every text change, once to get the matches, and another time to get the replacements. I think the performance issues come from the `rg-match` module running as well. I think it might make sense to chache the `rg` results in `state`, and have the incremental preview as well as `rg-match` access that cache? That would reduce the number of times `rg` needs to be run.

chrisgrieser commented on lua/rip-substitute/config.lua:
```diff
@@ -20,14 +20,23 @@ local defaultConfig = {
 	keymaps = {
 		-- normal & visual mode
 		confirm = "<CR>",
+		confirmAll = "<S-CR>",
```
for the sake of not introducing a breaking change for existing users, the keymaps should be switched.

`confirm` should be used to confirm all, and your feature should introduce a `confirmSingle` which uses a new keymaps `<S-CR>`. Yes, it's not ergonomic, but otherwise your PR creates a breaking change which will irrirate users.

chrisgrieser commented on lua/rip-substitute/config.lua:
```diff
@@ -76,9 +85,9 @@ function M.setup(userConfig)
-		local msg =
-			"`incrementalPreview.hlGroups.{name}` is deprecated, use `incrementalPreview.matchHlGroup` instead."
-		notify(msg, "warn")
+		-- local msg =
```
as annoying as it is, I'm afraid we will need a deprecation notice for the now basically reverted setting…

stefanwatt replied:
just to make sure I understand correctly:
the new deprecation notice would be:
	"`incrementalPreview.matchHlGroup` is deprecated, use `incrementalPreview.hlGroups.{name}` instead."
	
	?

chrisgrieser replied:
yep

chrisgrieser commented on lua/rip-substitute/config.lua:
```diff
@@ -20,14 +20,23 @@ local defaultConfig = {
 		abort = "q",
-		prevSubst = "<Up>",
-		nextSubst = "<Down>",
+		prevSubst = "<A-Up>",
+		nextSubst = "<A-Down>",
+		prevMatch = "<Up>",
+		nextMatch = "<Down>",
```
same thing as with `confirm`, this one creates a breaking change for existing users, and should therefore be avoided.

I think for `prevMatch` and `nextMatch`, using `n` and `N` might actually be a good fit? It fits with how vim users expect the keys to work, it prevents the need of changing the `nextSubst` keymaps (and thus avoids a breaking change), and the default `n/N` are not really needed in the tiny popup window anyway.

stefanwatt replied:
Done 👍

chrisgrieser commented on lua/rip-substitute/matches.lua:
```diff
@@ -0,0 +1,221 @@
+	end
+end
+
+function M.selectPrevMatch()
```
selectPrevMatch and selectNextMatch could be merged into one function taking the direction as a parameter

stefanwatt replied:
are you ok with keeping shorthand functions like this:
``` lua
---@param direction "next" | "prev"
local function cycleMatches(direction)
...
function M.selectNextMatch()
	cycleMatches("next")
end

function M.selectPrevMatch()
	cycleMatches("prev")
end
```

chrisgrieser replied:
why even the wrapper functions? You can just assign `cycleMatches` to the keymap definitions for the popup win?

stefanwatt replied:
perhaps I'm ignorant here...
As far as I'm aware the only way to do that would be with an inline function since `vim.keymap.set` doesn't have a mechanism to pass extra args to the function.
So If i have to write function()...end, I might as well do it like this instead of the inline function.
That was my thinking process anyway

chrisgrieser commented on lua/rip-substitute/utils.lua:
```diff
@@ -1,6 +1,26 @@
+
+---@param table table
+---@param cb function(value: any): boolean
+function M.index_of(table, cb)
```
If this case helper function is only used in the `matches` module, I think helper functions like these can simply be moved to that module?

(btw, not camelCase)

stefanwatt replied:
yeah, I guess.
I didn't know vim already had a lot of these higher order table functions etc. so I used to always have them in a utils file when developing a plugin.

I put it into matches and also made it less generic and therefore more concise.

chrisgrieser commented on lua/rip-substitute/matches.lua:
```diff
@@ -0,0 +1,221 @@
+end
+
+---@param matches RipSubstituteMatch[]
+---@return RipSubstituteMatch | nil, string |nil
```
just a minor thing, but here you have two returns in one line, while further above, you have two returns in two lines. For the sake of consistency, I think one return per line would be better

chrisgrieser commented on lua/rip-substitute/matches.lua:
```diff
@@ -0,0 +1,221 @@
+
+---@param matches RipSubstituteMatch[]
+---@return RipSubstituteMatch | nil, string |nil
+function M.getClosestMatchAfterCursor(matches)
```
a lot of variables in this function are snake_cased

chrisgrieser commented on lua/rip-substitute/matches.lua:
```diff
@@ -0,0 +1,221 @@
+	local code, matched = rg.runRipgrep {
+		"--line-number",
+		"--column",
+		"--vimgrep",
```
same issue as in the other comment `--vimgrep` is probably unneeded

chrisgrieser commented on lua/rip-substitute/matches.lua:
```diff
@@ -0,0 +1,221 @@
+		)
+	end
+	state.selectedMatch = state.matches[newSelectedMatchIndex]
+	--TODO: this will probably clear too many highlights
```
yes, an issue will be multiple matches in one line, for instance. Not totally sure what would be the best solution here though

chrisgrieser commented on lua/rip-substitute/rg-operations.lua:
```diff
@@ -91,113 +115,131 @@ function M.executeSubstitution()
+---@param matches RipSubstituteMatch
+---@param viewStartLnum number
+---@param viewEndLnum number
+---@return number|nil, number | nil
```
should be 2 return lines

stefanwatt replied:
done 👍

chrisgrieser commented on lua/rip-substitute/rg-operations.lua:
```diff
@@ -91,113 +115,131 @@ function M.executeSubstitution()
+---@param selected boolean
+---@param targetBuf number
+---@param incPreviewNs number
+function M.highlightReplacement(match, selected, targetBuf, incPreviewNs)
```
`targetBuf` and `incPreviewNs` do not need to be passed, since they can be accessed via `state`.

chrisgrieser commented on lua/rip-substitute/rg-operations.lua:
```diff
@@ -91,113 +115,131 @@ function M.executeSubstitution()
+--
+---@param match RipSubstituteMatch
+---@param selected boolean
+---@param targetBuf number
+---@param incPreviewNs number
```
`targetBuf` and `incPreviewNs` do not need to be passed, since they can be accessed via `state`.

chrisgrieser commented on lua/rip-substitute/popup-win.lua:
```diff
@@ -214,6 +221,7 @@ end
 ---@param searchPrefill string
 function M.openSubstitutionPopup(searchPrefill)
 	local rg = require("rip-substitute.rg-operations")
+	local rgMatches = require("rip-substitute.matches")
```
you are importing this twice, once in the function and once at the top of the file

stefanwatt replied:
oops
removed the unnecessary one

chrisgrieser reviewed (changes requested):
a bunch of more comments, some minor things I noticed, but also some more important stuff like on the keymaps.

stefanwatt commented:
Wow. That's a lot to take in 😄

I think I have a better idea of what you are looking for now.

I agree with all the remarks and will try to adapt the PR accordingly.

> Sorry, if I am came off as nitpicky or something

All good 👍 To be honest I was getting a bit into my feelings because of some of the critique, but that might just be me. I'm a sucker for words of affirmation 😄
I know it's also a lot of work for you to maintain your projects.

> imho keeping noise to a minimum is essential in large PRs

completely fair. I guess I got too excited and pushed too early.

stefanwatt commented:
Ok I think I took all your remarks into account and made adaptations accordingly.
Now I'll try to separate out the preparatory changes.
